### PR TITLE
Test coverage for GOAWAY on mid-stream headers 

### PR DESCRIPTION
**Pull request by johanandren**
References #3268

Actual fix was implemented in trailing headers work (I think)

### CODE REVIEW

**Comment by johanandren** [comment]
Now I actually noticed a `reject mid-stream HEADERS with endStream = false` test case in Http2ServerSpec so perhaps at least the server side was already covered?

**Comment by jrudolph** [comment]
> Now I actually noticed a `reject mid-stream HEADERS with endStream = false` test case in Http2ServerSpec so perhaps at least the server side was already covered?

Yes, looked into that again, it seems that the test which I found weird, didn't fail because it never pulled the network to see the GOAWAY and we never close or fail the connection itself properly.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala [hunk 1]
```diff
@@ -336,6 +336,13 @@ class Http2ServerSpec extends AkkaSpecWithMaterializer("""
         // Wait to give the warning (that we hope not to see) time to pop up.
         Thread.sleep(100)
       }
+      "fail if headers are sent mid request stream" inAssertAllStagesStopped new WaitingForRequestData {
```
Comment (jrudolph) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ServerSpec.scala:339]: Maybe we can remove that one and keep the client one.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala [hunk 2]
```diff
@@ -126,6 +126,19 @@ class Http2ClientSpec extends AkkaSpecWithMaterializer("""
...
       }
 
+      "GOAWAY when the response has headers mid stream" in new SimpleRequestResponseRoundtripSetup {
+        val streamId = 0x1
+        emitRequest(streamId, HttpRequest(uri = "http://www.example.com/"))
+        expect[HeadersFrame]()
+
+        sendDATA(streamId, endStream = false, ByteString("wowdata"))
+        val headerBlock = encodeHeaders(Seq(RawHeader("X-Mid-Stream", "Wowsuchmid")))
+        sendFrame(HeadersFrame(streamId, endStream = false, endHeaders = true, headerBlock, None))
```
Comment (jrudolph) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala:136]: Why would that be the second one? Isn't that interpreted as the first HEADERS of the response?

Comment (johanandren) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala:136]: Yeah, I don't know what I was thinking here TBH, this doesn't test the right thing at all.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala [hunk 3]
```diff
@@ -126,6 +127,29 @@ class Http2ClientSpec extends AkkaSpecWithMaterializer("""
...
+        response.entity shouldBe a[Chunked]
+
+        sendDATA(streamId, endStream = false, ByteString("asdf"))
+        sendHEADERS(streamId, endStream = false, Seq(RawHeader("X-mid-stream", "badvalue")))
+
+        // FIXME need something like the Http2ServerSpec.expectWindowUpdate() here
+        expectFrameFlagsStreamIdAndPayload(FrameType.WINDOW_UPDATE)
+        
+        val (_, error) = expectGOAWAY(streamId)
+        error should ===(ErrorCode.PROTOCOL_ERROR)
```
Comment (johanandren) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala:150]: Also manually verified by failing this test and seeing that there is a debug text in the log mentioning mid stream headers in addition to the PROTOCOL_ERROR

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala [hunk 4]
```diff
@@ -126,6 +127,29 @@ class Http2ClientSpec extends AkkaSpecWithMaterializer("""
...
+        ))
+
+        val response = expectResponse()
+        response.entity shouldBe a[Chunked]
+
+        sendDATA(streamId, endStream = false, ByteString("asdf"))
+        sendHEADERS(streamId, endStream = false, Seq(RawHeader("X-mid-stream", "badvalue")))
+
+        // FIXME need something like the Http2ServerSpec.expectWindowUpdate() here
+        expectFrameFlagsStreamIdAndPayload(FrameType.WINDOW_UPDATE)
```
Comment (johanandren) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala:147]: Is there a better way to do this (less fragile)?

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala (hunk 3 above)
Comment (jrudolph) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala:150]: You could also use `expect[GoAwayFrame]()` if you want to assert on the debug string.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala (hunk 4 above)
Comment (jrudolph) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala:147]: It's triggered by `sendDATA` so you could at least pull it up a line. The infrastructure in Http2ServerSpec (`pollForWindowUpdates`) would probably also work here but we haven't pulled it out so far.

**Review by jrudolph — APPROVED** [review]
LGTM

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala (hunk 3 above)
Comment (johanandren) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala:150]: Ah, nice, that's good to make sure the test isn't accidentally failing from abusing the protocol in some other way, I'll change.

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala (hunk 4 above)
Comment (johanandren) [reply akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala]: I'll leave it like this for now (but pulled it up next to the sendDATA).

File: akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala [hunk 5]
```diff
@@ -126,6 +127,29 @@ class Http2ClientSpec extends AkkaSpecWithMaterializer("""
...
+        response.entity shouldBe a[Chunked]
+
+        sendDATA(streamId, endStream = false, ByteString("asdf"))
+        expectFrameFlagsStreamIdAndPayload(FrameType.WINDOW_UPDATE)
+
+        sendHEADERS(streamId, endStream = false, Seq(RawHeader("X-mid-stream", "badvalue")))
+
+        val goAwayFrame = expect[GoAwayFrame]()
+        goAwayFrame.errorCode should ===(ErrorCode.PROTOCOL_ERROR)
+        goAwayFrame.debug.utf8String should ===("Got unexpected mid-stream HEADERS frame")
```
Comment (jrudolph) [inline akka-http2-support/src/test/scala/akka/http/impl/engine/http2/Http2ClientSpec.scala:150]: :+1:

**Review by jrudolph — APPROVED** [review]
LGTM

**Review by raboof — APPROVED** [review]

**Pull request merged**